styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -608,14 +611,21 @@ synchronized void tryPollPods() {
         continue;
       }
       final RunState runState = activeStates._2.get(workflowInstance.get());
+      final boolean deletePod;
       if (runState != null && isPodRunState(pod, runState)) {
         emitPodEvents(pod, runState);
-        cleanupWithRunState(workflowInstance.get(), pod, runState);
+        deletePod = cleanupWithRunState(workflowInstance.get(), pod, runState);
       } else {
         // The pod is stale as we fetched the active states _after_ listing all pods.
-        cleanupWithoutRunState(workflowInstance.get(), pod);
+        deletePod = cleanupWithoutRunState(workflowInstance.get(), pod);
+      }
+      if (deletePod) {
+        podsToDelete.add(pod.getMetadata().getName());
       }
     }
+
+    // Do blocking pod deletion after emitting events in order to avoid run states going stale
```

Comment:
Sorry @danielnorberg, I couldn't understand this comment. 
Can you explain why delaying the pod deletes to the end after the emitting the events avoid run states going stale?

Reply:
There is a comment about this above, is it sufficient or do we need additional documentation?

In general, when attempting to persist a run state change we rely on the `counter` to catch when the state has already been changed e.g. by another scheduler. So having read a run state from datastore, the likelihood that the state will have been changed (and the local copy is stale) increases as time passes. Therefore it is preferable to act on the run state as soon as possibly after it has been read.

If haven't  already, we should document this mechanism in more detail, but that is outside the scope of this PR.

Reply:
Got it